userns.go

```diff
@@ -413,7 +415,20 @@ func (s *store) getAutoUserNS(id string, options *types.AutoUserNsOptions, image
 	// mapping
 	usedUIDs = append(usedUIDs, options.AdditionalUIDMappings...)
 	usedGIDs = append(usedGIDs, options.AdditionalGIDMappings...)
-	availableUIDs, availableGIDs, err = findAvailableRange(size, size, availableUIDs, availableGIDs, usedUIDs, usedGIDs)
+
+	// To find available IDs, we must first subtract from the requested size the
```

Comment:
If I read the code above correctly, usedUIDs and usedGIDs are taken from existing containers. How other containers should affect the number of ids allocated for this container?

Reply:
the issue happens because when we subtract (subtractAll with host=false) the IDs already pre-allocated inside the container the HostID is also incremented:

```
{ContainerID:0 HostID:296608 Size:1}  <- this was preallocated, through options.AdditionalUIDMappings
{ContainerID:1 HostID:296610 Size:65535}  <- when we subtract the preallocated range the HostID that was previously 296609 is also changed.
```

The number of IDs is unchanged, but it is preferable to use `{ContainerID:1 HostID:296609 Size:65535}` so we don't leave a hole in the available IDs

Reply:
Not sure if I understand the issues correctly, but I have a simpler version of findAvailableIDRange (#856) that passes the unit tests in this commit. Current implementation seems overly complicated. Please let me know if #856 correctly addresses the issues.

Reply:
Thanks again for the review and apologies for the late reply. Did not mean to break anything on the PR request, which is why I asked about integration tests to vet it (given that the unit tests were passing). Let me look into PR #856.

Reply:
Looks like PR #856 is not a proper fix either (see my comments there where I described the mapping under 3 scenarios and why some look incorrect). 

For completeness, here are the same scenarios with this current PR (#853): 

1) "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :

{ContainerID:0 HostID:296608 Size:1}
{ContainerID:1 HostID:296609 Size:65535}    <<< LOOKS GOOD

2) Same as before, but change "run_as_user: 296609":

{ContainerID:0 HostID:296609 Size:1}
{ContainerID:1 HostID:296610 Size:65534}     <<< LOOKS GOOD

3) Same as before, but change "run_as_user: 296610":

{ContainerID:0 HostID:296610 Size:1}
{ContainerID:1 HostID:296609 Size:1}            <<< ERROR: HOST-ID DISCONTINUITY
{ContainerID:2 HostID:296611 Size:65533}

This last error was not addressed by this PR. It's a legal mapping, but will create problems for the lower-level runtimes since the mapping of host-IDs is not contiguous. 

Ideally all mappings would be a continuous range of container IDs starting at 0 mapped to a continues range of host IDs starting at whatever host ID is available for a given range size. For example, the last mapping should be {ContainerID: 0 HostID: 296610 Size: 65536}.

Reply:
> {ContainerID:0 HostID:296610 Size:1}
> {ContainerID:1 HostID:296609 Size:1} <<< ERROR: HOST-ID DISCONTINUITY
> {ContainerID:2 HostID:296611 Size:65533}

Why is there discontinuity in this case?

Reply:
> Why is there discontinuity in this case?

There is discontinuity in the sense that Container ID 0 maps to HostID 296610, so one expects Container ID 1 to map to 296611, and so on. But Container ID 1 maps to 296609 instead. To be more precise, the mapping is continuous but not linear. This is legal, but unnecessary and will likely cause all sorts of problems for lower-level runtimes.

Reply:
I think in this case it is necessary and it cannot be linear.

If you specify `run_as_user: 296610` then that is mapped to root in the container, but since 296609 is available, then it is also mapped inside the container to the first unused ID.

Reply:
Thanks, got it. Non-linearity in the container ID -> host ID mappings will cause extra complexity though. For example, in the Sysbox runtime we use a filesystem called shiftfs, which only works with linear mappings. I've not looked into the recent fs ID mapped mounts mechanism yet, so I don't know if that's affected. 

Also, even if we keep the non-linear mappings, CRI-O should have chosen to map container ID 1 to host ID 296608 (rather than 296609) since that was the first available host ID.